include/seastar/core/iostream-impl.hh

```diff
@@ -151,27 +152,31 @@ future<> output_stream<CharType>::write(temporary_buffer<CharType> p) noexcept {
 
 template <typename CharType>
 future<temporary_buffer<CharType>>
-input_stream<CharType>::read_exactly_part(size_t n, tmp_buf out, size_t completed) noexcept {
-    if (available()) {
-        auto now = std::min(n - completed, available());
-        std::copy(_buf.get(), _buf.get() + now, out.get_write() + completed);
-        _buf.trim_front(now);
-        completed += now;
-    }
-    if (completed == n) {
-        return make_ready_future<tmp_buf>(std::move(out));
-    }
+input_stream<CharType>::read_exactly_part(size_t n) noexcept {
+    temporary_buffer<CharType> out(n);
+    size_t completed{0U};
+    while (completed < n) {
+        size_t avail = available();
+        if (avail) {
+            auto now = std::min(n - completed, avail);
+            std::copy_n(_buf.get(), now, out.get_write() + completed);
+            _buf.trim_front(now);
+            completed += now;
+            if (completed == n) {
```

Comment:
note that I moved this `if (completed == n)` branch inside the `if (avail)` branch